okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,245 @@
+            public Void visit(QosException.RetryOther exception) {
+                if (maxNumRelocations <= 0) {
+                    callback.onFailure(call,
+                            new IOException("Exceeded the maximum number of allowed "
+                                    + "redirects for initial URL: " + call.request().url()));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
+                            exception.getRedirectTo().toString());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call,
+                                new IOException("Failed to determine valid redirect URL "
+                                        + "for '" + exception.getRedirectTo() + "' and base URLs "
+                                        + urls.getBaseUrls()));
+                    } else {
+                        Request redirectedRequest = request().newBuilder()
+                                .url(redirectTo.get())
+                                .build();
+                        client.newCallWithMutableState(
+                                redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                .enqueue(callback);
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public Void visit(QosException.Unavailable exception) {
+                Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                if (!backoff.isPresent()) {
+                    log.debug("Max number of retries exceeded, failing call");
+                    callback.onFailure(call,
+                            new IOException("Failed to complete the request due to a "
+                                    + "server-side QoS condition: 503", exception));
+                } else {
+                    log.debug("Rescheduling call after backoff",
+                            SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                    scheduleExecution(callback, backoff.get());
+                }
+                return null;
+            }
+        };
+    }
+
+    @Override
+    public Response execute() throws IOException {
+        SettableFuture<Response> future = SettableFuture.create();
+        enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                future.setException(exception);
+            }
+
+            @Override
+            public void onResponse(Call call, Response response) throws IOException {
+                future.set(response);
+            }
+        });
+
+        try {
+            return future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Call was interrupted during execution");
```

Comment:
what about an `InterruptedIOException`?

Reply:
maybe. part of the other TODO, ok?

Reply:
Not sure why we'd want to defer a one line change - are there downsides?

Reply:
ah, you mean the existing one, sure.

Reply:
done